### PR TITLE
Edit metadata

### PR DESCRIPTION
**Pull request by ChrisHuynh333**
## What does this PR do and why?
In this PR, the edit action has been added to the metadata listing ui. Users with access level >= Maintainer level will have access. The edit action will open a modal where users can edit the metadata key and/or value.

The metadata/edit_fields controller will initially receive the edit form param and validate it before calling the Samples::Metadata::UpdateService.

Note: The PR and changes take into account add and delete metadata which will be added afterwards. 
- `services/samples/metadata/fields` folder currently only contains `edit_service` but an `add_service` will be added.
- The table action ellipsis only contains edit (or nothing for analysis metadata), but delete will be added to the dropdown.

## Screenshots or screen recordings
New edit action:
image

Edit modal:
image

Successful update:
image

Unchanged metadata error:
image

Key already exists error:
image

## How to set up and validate locally
1. Set up a sample with metadata, preferably with fields from both a user and analysis source.
2. Navigate to the sample metadata page
3. Edit both keys and values to ensure they update accordingly.
4. Ensure two keys with the same name cannot exist (including different cases/capitilization).
5. Ensure any metadata added by an analysis does not have the edit action available in the action dropdown.
6. Ensure users with access < Maintainer do not see the action column or action dropdown ellipsis.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [x] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

**Comment by ChrisHuynh333** [comment]
I think my process to figure out which field was edited is quite 'expensive' currently. All metadata gets passed to the back end, and I'm looping through the metadata until the edited field is found. I imagine there's probably a solution where we can isolate the edited field on the front end and only submit that, but figured I'd get reviewed and thoughts before pursuing.

File: app/controllers/projects/samples/metadata/fields_controller.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+        # Validates metadata edit params and builds the expected update_service param prior to calling
+        # MetadataController and the metadata update_service
+
+        # Param is received as:
+        # params: {sample: {edit_field: {key: {old_key: new_key}, value: {old_value: new_value}}}
+        # Fields that have not been changed will have equal old and new
+        def update # rubocop:disable Metrics/AbcSize
+          authorize! @project, to: :update_sample?
+          metadata_update_params = ::Samples::Metadata::Fields::EditService.new(@project, @sample, current_user,
```
Comment (ericenns) [inline app/controllers/projects/samples/metadata/fields_controller.rb:18]: **nitpick (blocking):** Can you update the `EditService` to actually edit the fields, so after validation etc it should call the `Samples::Metadata::UpdateService`

File: app/controllers/projects/samples/metadata/fields_controller.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                                                                                edit_field_params).execute
+          if metadata_update_params.empty?
+            render status: :unprocessable_entity,
+                   locals: { key: edit_field_params['edit_field']['key'].values[0],
+                             value: edit_field_params['edit_field']['value'].values[0] }
+          else
+            request[:sample] = metadata_update_params
+            Projects::Samples::MetadataController.dispatch(:update, request, response)
```
Comment (ericenns) [inline app/controllers/projects/samples/metadata/fields_controller.rb:26]: **nitpick (blocking):** As per comment above we don't wan't to call the other controller from this controller method.

File: app/services/samples/metadata/fields/edit_service.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        end
+
+        def execute
+          authorize! @project, to: :update_sample?
+
+          validate_sample_in_project
+
+          validate_edit_fields
+
+          construct_metadata_update_params
```
Comment (ericenns) [inline app/services/samples/metadata/fields/edit_service.rb:28]: **nitpick (blocking):** after crafting the update params this service should call the `Samples::Metadata::UpdateService`

**Review by ericenns — CHANGES_REQUESTED** [review]
Looking better, but still would like some more refactor as per comments below

File: app/controllers/projects/samples/metadata/fields_controller.rb (hunk 2 above)
Comment (ChrisHuynh333) [reply app/controllers/projects/samples/metadata/fields_controller.rb]: Updated in 5df4af3ff5328969250bf3de5a14d6bde116f993

File: app/services/samples/metadata/fields/edit_service.rb (hunk 3 above)
Comment (ChrisHuynh333) [reply app/services/samples/metadata/fields/edit_service.rb:28]: Updated in 5df4af3ff5328969250bf3de5a14d6bde116f993

File: app/controllers/projects/samples/metadata/fields_controller.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            render_params = get_render_status_and_message(updated_metadata_field)
+            render status: render_params[:status], locals: { type: render_params[:message][:type],
+                                                             message: render_params[:message][:message] }
+          end
+        end
+
+        private
+
+        def edit_field_params
+          params.require(:sample).permit(edit_field: {})
```
Comment (ericenns) [inline app/controllers/projects/samples/metadata/fields_controller.rb:31]: **nitpick (blocking):** You should be able to update this params method to be specific, i.e. use the keys present in comment above `update` method. So just incorporating `key` and `value` as being a hash should provide more context and then if bad keys were given it wouldn't try to pass them.

**Review by ericenns — CHANGES_REQUESTED** [review]
This is looking a lot better, just one minor comment.

File: app/controllers/projects/samples/metadata/fields_controller.rb (hunk 4 above)
Comment (ChrisHuynh333) [reply app/controllers/projects/samples/metadata/fields_controller.rb]: Hopefully I didn't misinterpret what you meant - Added in a0e2712c46389d63c119a6251e2dab1bc003c610

File: app/views/projects/samples/metadata/_edit_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<%= form_with(model: @sample, url: namespace_project_sample_metadata_field_path(sample_id: @sample.id), method: :patch) do |form| %>
+  <%= turbo_frame_tag 'update-alert' %>
+  <div class="grid gap-4">
+  <%= form.fields_for :edit_field do |field| %>
+    <%= field.fields_for :key do |metadata_key| %>
+      <div class="form-field">
+        <%= metadata_key.label t("projects.samples.show.metadata.edit.key"),
+                  class:
+                    "block mb-2 text-sm font-medium text-slate-900 dark:text-white" %>
+        <%= metadata_key.text_field key,
```
Comment (joshsadam) [inline app/views/projects/samples/metadata/_edit_form.html.erb:10]: **suggestion:** Can you try to set autofocus on this when the modal opens

File: app/views/projects/samples/metadata/_edit_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      </div>
+    <% end %>
+     <%= field.fields_for :value do |metadata_value| %>
+      <div class="form-field">
+        <%= metadata_value.label t("projects.samples.show.metadata.edit.value"),
+                  class:
+                    "block mb-2 text-sm font-medium text-slate-900 dark:text-white" %>
+        <%= metadata_value.text_field value,
+                        value: value,
+                        required: true,
```
Comment (joshsadam) [inline app/views/projects/samples/metadata/_edit_form.html.erb:24]: **question:** @ericenns how do you feel about a user wanting to remove a value from this, i.e. clearing the input?  Or are we just going to have a delete option?

**Review by joshsadam — COMMENTED** [review]
This looks really good to me.  I just have a couple questions.

File: app/views/projects/samples/metadata/_edit_form.html.erb (hunk 5 above)
Comment (ChrisHuynh333) [reply app/views/projects/samples/metadata/_edit_form.html.erb:10]: Fixed in f4e48c28724d8724dd996660540c38875bf81aef

File: config/locales/en.yml [hunk 7]
```diff
@@ -875,6 +884,11 @@ en:
           error: "The following list of samples failed to transfer:"
           no_samples_transferred_error: "Samples were not transferred for the following reasons:"
       metadata:
+        edit:
+          edit: Edit
```
Comment (joshsadam) [inline config/locales/en.yml:888]: **suggestion:** Can you change the label of the button to open the modal, and the button to save the updated metadata to `Update`.

Comment (ChrisHuynh333) [reply config/locales/en.yml:888]: Updated in 37288a651a4b8797332102de5ea22a1996b006d4

File: app/services/samples/metadata/fields/update_service.rb [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            end
+
+            @metadata_update_params['metadata'][@key_update.keys[0]] = ''
+          end
+
+          @metadata_update_params['metadata'][@key_update.values[0]] = @value_update.values[0]
+        end
+
+        # Checks if the new key already exists within the @sample.metadata
+        def validate_new_key
```
Comment (deepsidhu85) [inline app/services/samples/metadata/fields/update_service.rb:70]: I don't think this method and loop is required. You should be able to just call `@sample.metadata.key?(@key_update.values[0].downcase)` above in the `construct_metadata_update_params` method

File: test/controllers/projects/samples/metadata/fields/fields_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+          test 'cannot update sample if not a member with access to the project' do
+            sign_in users(:david_doe)
+            patch namespace_project_sample_metadata_field_path(@namespace, @project29, @sample32),
+                  params: { 'sample' => { 'update_field' => {
+                    'key' => { 'metadatafield1' => 'metadatafield3' },
+                    'value' => { 'value1' => 'value3' }
+                  } }, format: :turbo_stream }
+            assert_response :unauthorized
+          end
+        end
```
Comment (deepsidhu85) [inline test/controllers/projects/samples/metadata/fields/fields_controller_test.rb:82]: Please add a test for a user not being able to update analysis provided metadata

File: test/services/samples/metadata/fields/update_service_test.rb [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+          params = { 'update_field' => { 'key' => { 'metadatafield1' => 'metadatafield2' },
+                                         'value' => { 'value1' => 'value1' } } }
+
+          Samples::Metadata::Fields::UpdateService.new(@project29, @sample32, @user, params).execute
+
+          assert @sample32.errors.full_messages.include?(
+            I18n.t('services.samples.metadata.update_fields.key_exists', key: 'metadatafield2')
+          )
+        end
+      end
```
Comment (deepsidhu85) [inline test/services/samples/metadata/fields/update_service_test.rb:154]: Please add a test for a user not being able to update analysis provided metadata

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
This looks great just have a couple of comments below

File: app/services/samples/metadata/fields/update_service.rb (hunk 8 above)
Comment (ChrisHuynh333) [reply app/services/samples/metadata/fields/update_service.rb]: Updated in 2a79e5b24b98f16b80cf676792daa31801249850

File: test/controllers/projects/samples/metadata/fields/fields_controller_test.rb (hunk 9 above)
Comment (ChrisHuynh333) [reply test/controllers/projects/samples/metadata/fields/fields_controller_test.rb:108]: Updated in 2a79e5b24b98f16b80cf676792daa31801249850

File: test/services/samples/metadata/fields/update_service_test.rb (hunk 10 above)
Comment (ChrisHuynh333) [reply test/services/samples/metadata/fields/update_service_test.rb:216]: Updated in 2a79e5b24b98f16b80cf676792daa31801249850

**Review by deepsidhu85 — APPROVED** [review]
Looks good to me!

**Pull request merged**